File: backend/gym-v2/src/main/java/com/gym/v2/measurement/entity/Measurement.java
package com.gym.v2.measurement.entity;

import com.gym.v2.auth.entity.AppUser;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.Instant;

/**
 * GYMAPP-V2 Ölçüm Entity'si. Kullanıcıların kilo, yağ oranı ve bölgesel vücut ölçülerini
 * tarih bazlı olarak kaydeder. Her kayıt bir "anlık görüntü" (snapshot) niteliğindedir ve
 * gelişim grafiklerinin veri kaynağını oluşturur.
 *
 * Pure Java Politikası: Lombok kullanılmaz, tüm getter/setter'lar manueldir.
 */
@Entity
@Table(name = "measurements")
public class Measurement {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	// Ölçümın sahibi olan kullanıcı
	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "user_id", nullable = false)
	private AppUser user;

	// --- Temel Metrikler ---

	@Column(precision = 5, scale = 2)
	private BigDecimal weight; // Kilo (kg)

	@Column(precision = 5, scale = 2)
	private BigDecimal height; // Boy (cm)

	@Column(name = "body_fat_pct", precision = 4, scale = 1)
	private BigDecimal bodyFatPct; // Yağ oranı (%)

	@Column(name = "muscle_mass", precision = 5, scale = 2)
	private BigDecimal muscleMass; // Kas kütlesi (kg)

	// --- Bölgesel Ölçüler (cm) ---

	@Column(precision = 5, scale = 1)
	private BigDecimal chest; // Göğüs çevresi

	@Column(precision = 5, scale = 1)
	private BigDecimal waist; // Bel çevresi

	@Column(precision = 5, scale = 1)
	private BigDecimal shoulders; // Omuz çevresi

	@Column(name = "left_arm", precision = 5, scale = 1)
	private BigDecimal leftArm; // Sol kol çevresi

	@Column(name = "right_arm", precision = 5, scale = 1)
	private BigDecimal rightArm; // Sağ kol çevresi

	@Column(name = "left_leg", precision = 5, scale = 1)
	private BigDecimal leftLeg; // Sol bacak çevresi

	@Column(name = "right_leg", precision = 5, scale = 1)
	private BigDecimal rightLeg; // Sağ bacak çevresi

	@Column(precision = 5, scale = 1)
	private BigDecimal hips; // Kalça çevresi

	// --- Ek Bilgiler ---

	@Column(length = 500)
	private String notes; // Kullanıcı notları

	@Column(name = "created_at", nullable = false, updatable = false, columnDefinition = "TIMESTAMPTZ")
	private Instant createdAt;

	// Antrenörle paylaşılma durumu
	@Column(name = "is_shared_with_coach", nullable = false)
	private Boolean isSharedWithCoach = false;

	@PrePersist
	protected void onCreate() {
		this.createdAt = Instant.now();
	}

	// ===================== GETTER & SETTER =====================

	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public AppUser getUser() {
		return user;
	}

	public void setUser(AppUser user) {
		this.user = user;
	}

	public BigDecimal getWeight() {
		return weight;
	}

	public void setWeight(BigDecimal weight) {
		this.weight = weight;
	}

	public BigDecimal getHeight() {
		return height;
	}

	public void setHeight(BigDecimal height) {
		this.height = height;
	}

	public BigDecimal getBodyFatPct() {
		return bodyFatPct;
	}

	public void setBodyFatPct(BigDecimal bodyFatPct) {
		this.bodyFatPct = bodyFatPct;
	}

	public BigDecimal getMuscleMass() {
		return muscleMass;
	}

	public void setMuscleMass(BigDecimal muscleMass) {
		this.muscleMass = muscleMass;
	}

	public BigDecimal getChest() {
		return chest;
	}

	public void setChest(BigDecimal chest) {
		this.chest = chest;
	}

	public BigDecimal getWaist() {
		return waist;
	}

	public void setWaist(BigDecimal waist) {
		this.waist = waist;
	}

	public BigDecimal getShoulders() {
		return shoulders;
	}

	public void setShoulders(BigDecimal shoulders) {
		this.shoulders = shoulders;
	}

	public BigDecimal getLeftArm() {
		return leftArm;
	}

	public void setLeftArm(BigDecimal leftArm) {
		this.leftArm = leftArm;
	}

	public BigDecimal getRightArm() {
		return rightArm;
	}

	public void setRightArm(BigDecimal rightArm) {
		this.rightArm = rightArm;
	}

	public BigDecimal getLeftLeg() {
		return leftLeg;
	}

	public void setLeftLeg(BigDecimal leftLeg) {
		this.leftLeg = leftLeg;
	}

	public BigDecimal getRightLeg() {
		return rightLeg;
	}

	public void setRightLeg(BigDecimal rightLeg) {
		this.rightLeg = rightLeg;
	}

	public BigDecimal getHips() {
		return hips;
	}

	public void setHips(BigDecimal hips) {
		this.hips = hips;
	}

	public String getNotes() {
		return notes;
	}

	public void setNotes(String notes) {
		this.notes = notes;
	}

	public Instant getCreatedAt() {
		return createdAt;
	}

	public Boolean getIsSharedWithCoach() {
		return isSharedWithCoach;
	}

	public void setIsSharedWithCoach(Boolean isSharedWithCoach) {
		this.isSharedWithCoach = isSharedWithCoach;
	}

}

File: frontend/gymapp-v2/lib/features/measurement/models/measurement.dart
/// GYMAPP-V2 Ölçüm Modeli.
/// Backend MeasurementDTO ile birebir eşleşir.
/// Tüm bölgesel ölçüler opsiyoneldir (nullable).
class Measurement {
  final int? id;
  final double? weight; // Kilo (kg)
  final double? height; // Boy (cm)
  final double? bodyFatPct; // Yağ oranı (%)
  final double? muscleMass; // Kas kütlesi (kg)
  final double? chest; // Göğüs çevresi (cm)
  final double? waist; // Bel çevresi (cm)
  final double? shoulders; // Omuz çevresi (cm)
  final double? leftArm; // Sol kol çevresi (cm)
  final double? rightArm; // Sağ kol çevresi (cm)
  final double? leftLeg; // Sol bacak çevresi (cm)
  final double? rightLeg; // Sağ bacak çevresi (cm)
  final double? hips; // Kalça çevresi (cm)
  final String? notes; // Kullanıcı notları
  final DateTime? createdAt; // Ölçüm tarihi
  final bool isSharedWithCoach; // Antrenörle paylaşılma durumu

  Measurement({
    this.id,
    this.weight,
    this.height,
    this.bodyFatPct,
    this.muscleMass,
    this.chest,
    this.waist,
    this.shoulders,
    this.leftArm,
    this.rightArm,
    this.leftLeg,
    this.rightLeg,
    this.hips,
    this.notes,
    this.createdAt,
    this.isSharedWithCoach = false,
  });

  /// JSON → Measurement dönüşümü (API'den gelen veri)
  factory Measurement.fromJson(Map<String, dynamic> json) {
    return Measurement(
      id: json['id'] as int?,
      weight: (json['weight'] as num?)?.toDouble(),
      height: (json['height'] as num?)?.toDouble(),
      bodyFatPct: (json['bodyFatPct'] as num?)?.toDouble(),
      muscleMass: (json['muscleMass'] as num?)?.toDouble(),
      chest: (json['chest'] as num?)?.toDouble(),
      waist: (json['waist'] as num?)?.toDouble(),
      shoulders: (json['shoulders'] as num?)?.toDouble(),
      leftArm: (json['leftArm'] as num?)?.toDouble(),
      rightArm: (json['rightArm'] as num?)?.toDouble(),
      leftLeg: (json['leftLeg'] as num?)?.toDouble(),
      rightLeg: (json['rightLeg'] as num?)?.toDouble(),
      hips: (json['hips'] as num?)?.toDouble(),
      notes: json['notes'] as String?,
      createdAt:
          json['createdAt'] != null ? DateTime.parse(json['createdAt'] as String) : null,
      isSharedWithCoach: json['isSharedWithCoach'] as bool? ?? false,
    );
  }

  /// Measurement → JSON dönüşümü (API'ye gönderilecek veri)
  Map<String, dynamic> toJson() {
    return {
      if (weight != null) 'weight': weight,
      if (height != null) 'height': height,
      if (bodyFatPct != null) 'bodyFatPct': bodyFatPct,
      if (muscleMass != null) 'muscleMass': muscleMass,
      if (chest != null) 'chest': chest,
      if (waist != null) 'waist': waist,
      if (shoulders != null) 'shoulders': shoulders,
      if (leftArm != null) 'leftArm': leftArm,
      if (rightArm != null) 'rightArm': rightArm,
      if (leftLeg != null) 'leftLeg': leftLeg,
      if (rightLeg != null) 'rightLeg': rightLeg,
      if (hips != null) 'hips': hips,
      if (notes != null) 'notes': notes,
    };
  }

  Measurement copyWith({
    int? id,
    double? weight,
    double? height,
    double? bodyFatPct,
    double? muscleMass,
    double? chest,
    double? waist,
    double? shoulders,
    double? leftArm,
    double? rightArm,
    double? leftLeg,
    double? rightLeg,
    double? hips,
    String? notes,
    DateTime? createdAt,
    bool? isSharedWithCoach,
  }) {
    return Measurement(
      id: id ?? this.id,
      weight: weight ?? this.weight,
      height: height ?? this.height,
      bodyFatPct: bodyFatPct ?? this.bodyFatPct,
      muscleMass: muscleMass ?? this.muscleMass,
      chest: chest ?? this.chest,
      waist: waist ?? this.waist,
      shoulders: shoulders ?? this.shoulders,
      leftArm: leftArm ?? this.leftArm,
      rightArm: rightArm ?? this.rightArm,
      leftLeg: leftLeg ?? this.leftLeg,
      rightLeg: rightLeg ?? this.rightLeg,
      hips: hips ?? this.hips,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
      isSharedWithCoach: isSharedWithCoach ?? this.isSharedWithCoach,
    );
  }
}
